src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/PropertyRef.cs

```diff
@@ -1,21 +1,113 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
 namespace System.Text.Json.Serialization.Metadata
 {
-    internal readonly struct PropertyRef
+    /// <summary>
+    /// Represents a UTF-8 encoded JSON property name and its associated <see cref="JsonPropertyInfo"/>, if available.
+    /// PropertyRefs use byte sequence equality, so equal JSON strings with alternate encodings or casings are not equal.
+    /// Used as a first-level cache for property lookups before falling back to UTF decoding and string comparison.
+    /// </summary>
+    internal readonly struct PropertyRef(ulong key, JsonPropertyInfo? info, byte[] utf8PropertyName) : IEquatable<PropertyRef>
     {
-        public PropertyRef(ulong key, JsonPropertyInfo info, byte[] nameFromJson)
+        // The length of the property name embedded in the key (in bytes).
+        // The key is a ulong (8 bytes) containing the first 7 bytes of the property name
+        // followed by a byte representing the length.
+        private const int PropertyNameKeyLength = 7;
+
+        /// <summary>
+        /// A custom hashcode produced from the UTF-8 encoded property name.
+        /// </summary>
+        public readonly ulong Key = key;
+
+        /// <summary>
+        /// The <see cref="JsonPropertyInfo"/> associated with the property name, if available.
+        /// </summary>
+        public readonly JsonPropertyInfo? Info = info;
+
+        /// <summary>
+        /// Caches a heap allocated copy of the UTF-8 encoded property name.
+        /// </summary>
+        public readonly byte[] Utf8PropertyName = utf8PropertyName;
+
+        public bool Equals(PropertyRef other) => Equals(other.Utf8PropertyName, other.Key);
+        public override bool Equals(object? obj) => obj is PropertyRef other && Equals(other);
+        public override int GetHashCode() => Key.GetHashCode();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Equals(ReadOnlySpan<byte> propertyName, ulong key)
         {
-            Key = key;
-            Info = info;
-            NameFromJson = nameFromJson;
+            if (key == Key)
+            {
+                // We compare the whole name, although we could skip the first 7 bytes (but it's not any faster)
+                if (propertyName.Length <= PropertyNameKeyLength ||
```

Comment:
I don't understand the logic here... why is any property name that's <= 7 characters considered equal?

Reply:
Oh, I see, the Key that was already checked represents up to the first 7 characters. Got it.